src/async_impl/h3_client/mod.rs

```diff
@@ -0,0 +1,147 @@
+#![cfg(feature = "http3")]
+
+mod pool;
+
+use crate::async_impl::h3_client::pool::{Key, Pool, PoolClient};
+use crate::error::{BoxError, Error, Kind};
+use crate::{error, Body};
+use futures_util::future;
+use h3_quinn::Connection;
+use http::{Request, Response};
+use hyper::Body as HyperBody;
+use log::debug;
+use std::future::Future;
+use std::net::{IpAddr, SocketAddr};
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+pub struct H3Builder {
+    pool_idle_timeout: Option<Duration>,
+    pool_max_idle_per_host: usize,
+    local_addr: Option<IpAddr>,
+}
+
+impl Default for H3Builder {
+    fn default() -> Self {
+        Self {
+            pool_idle_timeout: Some(Duration::from_secs(90)),
+            pool_max_idle_per_host: usize::MAX,
+            local_addr: None,
+        }
+    }
+}
+
+impl H3Builder {
+    pub fn build(self, tls: rustls::ClientConfig) -> H3Client {
+        let config = quinn::ClientConfig::new(Arc::new(tls));
+        let socket_addr = match self.local_addr {
+            Some(ip) => SocketAddr::new(ip, 0),
+            None => "[::]:0".parse::<SocketAddr>().unwrap(),
+        };
+
+        let mut endpoint =
+            quinn::Endpoint::client(socket_addr).expect("unable to create QUIC endpoint");
+        endpoint.set_default_client_config(config);
+
+        H3Client {
+            endpoint,
+            pool: Pool::new(self.pool_max_idle_per_host, self.pool_idle_timeout),
+        }
+    }
+
+    pub fn set_pool_idle_timeout(&mut self, timeout: Option<Duration>) {
+        self.pool_idle_timeout = timeout;
+    }
+
+    pub fn set_pool_max_idle_per_host(&mut self, max: usize) {
+        self.pool_max_idle_per_host = max;
+    }
+
+    pub fn set_local_addr(&mut self, addr: Option<IpAddr>) {
+        self.local_addr = addr;
+    }
+}
+
+#[derive(Clone)]
+pub struct H3Client {
+    endpoint: quinn::Endpoint,
+    pool: Pool,
+    // TODO: Since resolution is perform internally in Hyper,
```

Comment:
The resolver is specified in `src/connect.rs`. It's either `hyper::client::connect::dns::GaiResolver`, or one with overrides, or the `TrustDnsResolver`. You could perhaps save it in here roughly as a `Box<dyn Service<Uri, Response = WhatsTheAddrType>>`?